internal/postgres/users.go

```diff
@@ -162,5 +170,76 @@ SELECT pg_catalog.format('GRANT ALL PRIVILEGES ON DATABASE %I TO %I',
 
 	log.V(1).Info("wrote PostgreSQL users", "stdout", stdout, "stderr", stderr)
 
+	// The operator will attemtp to write schemas for the users in the spec if
+	// 	* the feature gate is enabled and
+	// 	* the cluster is annotated.
+	if util.DefaultMutableFeatureGate.Enabled(util.AutoCreateUserSchema) {
+		if _, annotationExists := cluster.Annotations[naming.AutoCreateUserSchemaAnnotation]; annotationExists {
+			log.V(1).Info("Writing schemas for users.")
+			err = WriteUsersSchemasInPostgreSQL(ctx, exec, users)
+		}
+	}
+
+	return err
+}
+
+// WriteUsersSchemasInPostgreSQL will create a schema for each user in each database that user has access to
+// NOTE: Does this mean if no dbs are specified, no schemas are created?
+func WriteUsersSchemasInPostgreSQL(ctx context.Context, exec Executor,
+	users []v1beta1.PostgresUserSpec) error {
+
+	log := logging.FromContext(ctx)
+
+	var err error
+	var stdout string
+	var stderr string
+
+	for i := range users {
+		spec := users[i]
+
+		if RESERVED_SCHEMA_NAMES[string(spec.Name)] {
+			log.V(1).Info("Skipping schema creation for user with reserved name",
+				"name", string(spec.Name))
+			continue
+		}
+
+		databases := []string{}
+		for _, db := range spec.Databases {
+			databases = append(databases, string(db))
+		}
+
+		if len(databases) == 0 {
+			continue
+		}
+
+		var sql bytes.Buffer
+
+		// Prevent unexpected dereferences by emptying "search_path". The "pg_catalog"
+		// schema is still searched, and only temporary objects can be created.
+		// - https://www.postgresql.org/docs/current/runtime-config-client.html#GUC-SEARCH-PATH
+		_, _ = sql.WriteString(`SET search_path TO '';`)
+
+		_, _ = sql.WriteString(`SELECT UNNEST(ARRAY[:databases]);`)
+
+		stdout, stderr, err = exec.ExecInDatabasesFromQuery(ctx,
+			sql.String(),
+			strings.Join([]string{
+				// Quiet NOTICE messages from IF EXISTS statements.
+				// - https://www.postgresql.org/docs/current/runtime-config-client.html
+				`SET client_min_messages = WARNING;`,
+
+				"CREATE SCHEMA IF NOT EXISTS AUTHORIZATION :username;",
```

Comment:
This creates a schema named :username, owned by :username.

Our only other schema that we lock down that I can see is `pgbouncer`, and we lock that with 

```sql
REVOKE ALL PRIVILEGES
    ON SCHEMA :"namespace" FROM PUBLIC, :"username";
 GRANT USAGE
    ON SCHEMA :"namespace" TO :"username";
```
--[cite](https://github.com/CrunchyData/postgres-operator/blob/master/internal/pgbouncer/postgres.go#L169-L172)

The difference being something like (rhino1 created with the `authorization` code, rhino2 granted/revoked as per other code):

```
postgres=# \dn+
                                       List of schemas
  Name  |       Owner       |           Access privileges            |      Description       
--------+-------------------+----------------------------------------+------------------------
 public | pg_database_owner | pg_database_owner=UC/pg_database_owner+| standard public schema
        |                   | =U/pg_database_owner                   | 
 rhino1 | rhino1            |                                        | 
 rhino2 | postgres          | postgres=UC/postgres                  +| 
        |                   | rhino2=U/postgres                      | 
(3 rows)
```

Reply:
These permissions LGTM.